libraries/AP_OSD/AP_OSD_MSP_DisplayPort.cpp

```diff
@@ -0,0 +1,119 @@
+ * See the GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License along
+ * with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+/*
+  OSD backend for MSP
+ */
+#include <AP_MSP/AP_MSP.h>
+#include <AP_MSP/msp.h>
+#include "AP_OSD_MSP_DisplayPort.h"
+
+static const struct AP_Param::defaults_table_struct defaults_table[] = {
+    /*
+    { "PARAM_NAME",       value_float }
+    */
+};
+
+extern const AP_HAL::HAL &hal;
+
+// initialise backend
+bool AP_OSD_MSP_DisplayPort::init(void)
+{
+#if HAL_MSP_ENABLED
+    // check if we have a DisplayPort backend to use
+    AP_MSP *msp = AP::msp();
+    if (msp == nullptr) {
+        gcs().send_text(MAV_SEVERITY_WARNING,"MSP backend not available");
+        return false;
+    }
+    _displayport = msp->find_protocol(AP_SerialManager::SerialProtocol_MSP_DisplayPort);
+    if (_displayport == nullptr) {
+        gcs().send_text(MAV_SEVERITY_WARNING,"MSP DisplayPort uart not available");
+        return false;
+    }
+    // re-init port here for use in this thread
+    _displayport->init_uart();
+    return true;
+#else
+    return false;
+#endif
+}
+
+void AP_OSD_MSP_DisplayPort::clear(void)
+{
+#if HAL_MSP_ENABLED
+    // clear remote MSP screen
+    _displayport->msp_displayport_clear_screen();
+#endif
+}
+
+void AP_OSD_MSP_DisplayPort::write(uint8_t x, uint8_t y, const char* text)
+{
+#if HAL_MSP_ENABLED
+    _displayport->msp_displayport_write_string(x, y, 0, text);
+#endif
+}
+
+void AP_OSD_MSP_DisplayPort::write(uint8_t x, uint8_t y, bool blink, const char *fmt, ...)
+{
+#if HAL_MSP_ENABLED
+    char buf[32+1]; // +1 for snprintf null-termination
+    va_list ap;
+    va_start(ap, fmt);
+    int res = hal.util->vsnprintf(buf, sizeof(buf), fmt, ap);
+    res = MIN(res, int(sizeof(buf)));
+    if (res < int(sizeof(buf))-1) {
+        _displayport->msp_displayport_write_string(x, y, blink, buf);
+    }
+    va_end(ap);
+#endif
+}
+
+void AP_OSD_MSP_DisplayPort::flush(void)
+{
+#if HAL_MSP_ENABLED
+    // grab the screen and force a redraw
+    _displayport->msp_displayport_grab();
+    _displayport->msp_displayport_draw_screen();
+
+    // ok done processing displayport data
+    // let's process incoming MSP frames (and reply if needed)
+    _displayport->process_incoming_data();
+#endif
+}
+
+// override built in positions with defaults for MSP OSD
+void AP_OSD_MSP_DisplayPort::setup_defaults(void)
+{
+    AP_Param::set_defaults_from_table(defaults_table, ARRAY_SIZE(defaults_table));
+}
+
+AP_OSD_Backend *AP_OSD_MSP_DisplayPort::probe(AP_OSD &osd)
+{
+    AP_OSD_MSP_DisplayPort *backend = new AP_OSD_MSP_DisplayPort(osd);
+    if (!backend) {
+        return nullptr;
+    }
+    if (!backend->init()) {
+        delete backend;
+        return nullptr;
+    }
+    return backend;
+}
+
+AP_OSD_MSP_DisplayPort::AP_OSD_MSP_DisplayPort(AP_OSD &osd):
+    AP_OSD_Backend(osd)
+{
+}
```

Comment:
Use the parent class constructor with `using AP_MSP_Backend::AP_MSP_Backend;` in the header instead

Reply:
ok